### ISSUE TITLE
regridding transformation to an identical axis, and transforming scalars and "normal" axes

### ISSUE THREAD

**Issue by AndrewWittenberg** [NOAA-PMEL/Ferret#1868]
As described at the end of in #1618, currently a regridding transformation to an identical axis is a no-op.  This is appropriate for many of the regridding transformations, but not all of them.  For example:

```
 	NOAA/PMEL TMAP
 	FERRET v7.4 (optimized)
 	Linux 2.6.32-696.23.1.el6.x86_64 64-bit - 04/10/18
 	10-Apr-18 19:02     

yes? def ax/x=1:4:1 xax
yes? let two = 2 + 0*x[gx=xax]
yes? list two[gx=xax@ngd]
             VARIABLE : 2 + 0*X[GX=XAX]
             SUBSET   : 4 points (X)
 1   / 1:  2.000
 2   / 2:  2.000
 3   / 3:  2.000
 4   / 4:  2.000
```

Here we've asked Ferret how many "good" points sit within each cell: the answer should be all 1's, not 2's.  Other examples of regridding transformations where Ferret actually needs to do something nontrivial, if the target axis matches the source axis:

`@ngd`, `@modngd`: Result should be 1 where valid, 0 elsewhere.
`@modnbd`: Result should be 0 where valid, 1 elsewhere.
`@var`, `@modvar`: Result should be 0 where valid, missing elsewhere.
`@nbin`: Result should be 1 where valid, missing elsewhere.
`@iin`: Result should be the weighted integral from the start of the target axis to each cell coordinate, just as if we were regridding to a non-identical axis.

(This desired behavior for `@iin` regridding is the main reason I'm submitting this ticket.)

A related issue is regular (non-regridding) transforms that are "normal" (orthogonal) to the axis of the source variable.  A trivial case is scalars, which are normal to _every_ axis:

```
yes? let a = 2
yes? list a[x=@ngd]
             VARIABLE : 2
          2.000
```

Here the answer should be 1.

Such transformations to consider: `@var`, `@shf`, `@ddc`, `@ddf`, `@ddb`, `@din`, `@iin`, `@ngd`, `@nbd`, `@loc`, `@weq`, `@cda`, `@cdb`, `@cia`, `@cib`, `@evn`, `@std`, `@wgt`.

For scalars, the results of many of these could simply be 1, 0, or the missing value.  Alternatively, Ferret could just give an error and say that transformations are disallowed for scalars.

For transformations normal to an existing axis:

```
yes? let a = {1,2,3,4}
yes? list a[x=@ngd, y=@ngd]
 **ERROR: illegal use of transformation:  not all dimensions present on grid
```

This seems acceptable -- Ferret just cops out.  (An alternative might be to simply ignore the y-oriented component of the multi-axis transformation.)  But then

```
yes? list a[y=@ngd]
             VARIABLE : {1,2,3,4}
             SUBSET   : 4 points (X)
 1   / 1:  1.000
 2   / 2:  2.000
 3   / 3:  3.000
 4   / 4:  4.000
```

This ought to either return all 1's, or give an error saying that such "normal" transformations are disallowed.

**Comment by AnsleyManke** [comment]
I've implemented all of the regridding to the same axis operations listed above. 

Leaving this open, as the rest of the comment about regular regridding operations isn't yet addressed.

**Comment by AnsleyManke** [comment]
This is done: transformations of scalars returns an error when the transform makes no sense. 

```
yes? let a = 2
yes? list a[x=@nbd]
             VARIABLE : 2 (# of points)
          0.0000
yes? list a[x=@ngd]
             VARIABLE : 2 (# of points)
          1.000

yes? list a[x=@var]
 **ERROR: illegal use of transformation: A[X=???@VAR,D=0] applied to point
yes? list a[x=@shf]
 **ERROR: illegal use of transformation: A[X=???@SHF:1,D=0] applied to point
yes? list a[x=@ddc]
 **ERROR: illegal use of transformation: A[X=???@DDC,D=0] applied to point
...
! Smoothers etc can just return the constant.
yes? list a[e=@sbx]
             VARIABLE : 2
          2.000

```

---------
Looking at the last example in the description above, with 

```
yes? let a = {1,2,3,4}
yes? list a[x=@ngd, y=@ngd]
 **ERROR: illegal use of transformation:  not all dimensions present on grid

yes? list a[y=@ngd]
             VARIABLE : {1,2,3,4}
             SUBSET   : 4 points (X)
 1   / 1:  1.000
 2   / 2:  2.000
 3   / 3:  3.000
 4   / 4:  4.000
```

This is in fact correct behavior. Some transformations, including @ave, @var, @ngd, @nbd, are applied in multiple directions as a single operation, computing the result over the whole block of data requested.  If the grid doesn't have all the directions requested for transformation, that's an error. The variable is on an x axis so the average only in the Y direction is meaningless and is ignored, and Ferret returns the data in X

Looking at another example, the result is I think what we're used to: can't do an XZ average on a surface variable, but asking for the variable averaged only in a direction orthogonal to the grid just ignores that averaging:

```
yes? use coads_climatology
yes? list sst[x=@ave,z=@ave]
 **ERROR: illegal use of transformation:  not all dimensions present on grid

yes? shade/l=1 sst[z=@ave]
```

Other transforms such as @SHF or @MAX are applied separately axis by axis, and asking for  a[x=@shf,y=@shf] just ignores a request in any direction that's not represented in the grid.

**Comment by AndrewWittenberg** [comment]
Thank you!  I hesitate to comment, since we're pretty far in the weeds now... but for the example

```
yes? list a[x=@ngd, y=@ngd]
 **ERROR: illegal use of transformation:  not all dimensions present on grid
```

wouldn't we want this to behave in a manner consistent with the "orthogonal" behavior, i.e. simply ignoring the y-transformation and proceeding with the 1-d transform along x, rather than issuing an error?

**Comment by AnsleyManke** [comment]
I agree on the logic of what you're saying here, and will have another look. It may be something that's a real issue in other contexts and we're just seeing a kind of odd example of it.

**Comment by AnsleyManke** [comment]
For the above last part of the a discussion, it's easily changed so that if the request is made for a multi-axis transform, but the grid doesn't contain more than 1 of the requested axes, we can just go do the 1-directional transform., and ignore the others that were requested.  Fix is in is_trans.F

```

yes? let a = {1,2,3,4}
yes? list a[x=@ngd, y=@ngd]
             VARIABLE : {1,2,3,4} (# of points)
             X        : 0.5 to 4.5 (number of valid)
          4.000

yes? use coads_climatology
 yes? list/y=5/L=3:7 sst[x=@ave,z=@ave]
             VARIABLE : SEA SURFACE TEMPERATURE (Deg C)
             FILENAME : coads_climatology.cdf
             FILEPATH : /home/users/tmap/ferret/linux/fer_dsets/data/
             SUBSET   : 5 points (TIME)
             LONGITUDE: 20E to 20E(380) (averaged)
             LATITUDE : 5N
 17-MAR      / 3:  28.00
 16-APR      / 4:  28.49
 16-MAY      / 5:  28.54
 16-JUN      / 6:  28.27
 16-JUL      / 7:  27.87
```

**Comment by AndrewWittenberg** [comment]
It looks like not all of these changes made it into the latest version I have -- in particular for scalars:

```
 	NOAA/PMEL TMAP
 	FERRET v7.41 (optimized)
 	Linux 2.6.32-754.el6.x86_64 64-bit - 07/09/18
 	10-Jul-18 11:56     

yes? let a = 2
yes? list a[x=@nbd]
             VARIABLE : 2
          2.000
yes? list a[x=@ngd]
             VARIABLE : 2
          2.000
```

This doesn't match the behavior you showed in your comment above (4 comments back from here).  The answers here should be 0 and 1.

**Comment by AnsleyManke** [comment]
yes, wrong version of flesh_out_axis.F was checked in.  In our last v7.41 version the above with @NGD and @NBD; also @SBX and @FAV returned a result. Now,

```
        NOAA/PMEL TMAP
        FERRET v7.42 (beta/debug)
        Linux 2.6.32-696.30.1.el6.x86_64 64-bit - 07/10/18
        10-Jul-18 10:30     

yes? ! Non-regridding transformations of scalars
yes? let a = 2
yes? list a[x=@nbd]
             VARIABLE : 2 (# of points)
          0.0000
yes? list a[x=@ngd]
             VARIABLE : 2 (# of points)
          1.000
yes? ! smoothers etc can return the constant's value
yes? list a[x=@sbx]
 **ERROR: illegal use of transformation: A[X=???@SBX:3,D=0] applied to point
yes? list a[x=@fav]
 **ERROR: illegal use of transformation: A[X=???@FAV:3,D=0] applied to point
```

**Comment by AndrewWittenberg** [comment]
Thanks for the fix in v7.42 (7/10/2018)!

My only final comment would be to make sure this behavior is documented -- in particular that any "normal" transformations are ignored.  This is intuitive for @ave, but maybe not for the rest.
